ansible/roles/gpu/tasks/nvidia-gpu-Flatcar.yaml

```
@@ -1,18 +1,197 @@
 ---
-- name: Load loop modules
-  modprobe:
-    name:
-      - loop
-      - i2c_core
-      - ipmi_msghandler
-    state: present
-
-# TODO: Replace me with proper ansible resources 
-- name: Load loop module
+- name: create /opt/lib directory
+  file:
+    path: "/opt/lib/"
+    state: directory
+
+- name: create /opt/var/log directory
+  file:
+    path: "/opt/var/log/"
+    state: directory
+
+# build kernel modules
+- name: load loop modules
+  command: modprobe -d / -a loop i2c_core ipmi_msghandler
+
+- name: install loop modules
   command: echo -e "loop\ni2c_core\nipmi_msghandler" | tee /etc/modules-load.d/driver.conf
 
-- name: Pull driver build image
-  command: docker pull nvcr.io/nvidia/driver:460.73.01-5.10.32-flatcar
+- name: pull driver build image
+  command: "docker pull nvcr.io/nvidia/driver:460.73.01-5.10.32-flatcar"
+
+# Note, this has be to run privileged as the driver module has to be loaded (briefly).
+- name: run the build
+  command: docker run -d --rm --privileged --pid=host -v /run/nvidia:/run/nvidia:shared -v /tmp/nvidia:/var/log -v /usr/lib64/modules:/usr/lib64/modules -v /opt:/opt --name nvidia-driver nvcr.io/nvidia/driver:460.73.01-5.10.32-flatcar
+
+- name: wait for build being done
+  command: "docker logs --tail 1 nvidia-driver"
+  register: build_log
+  retries: 150
+  delay: 10
+  until: "'Done, now waiting for signal' in build_log.stdout"
+
+# FIXME(tillt): consider a proper list of libs to grab - for that we would need a proper dependency tree!
+- name: tar nvidia-libs /usr/lib/x86_64-linux-gnu
+  command: docker exec -i nvidia-driver "/bin/sh" -c "tar czf /opt/lib/x86_64-linux-gnu.tar.gz /usr/lib/x86_64-linux-gnu"
+
+- name: tar nvidia-bins /usr/bin/nvidia*
+  command: docker exec -i nvidia-driver "/bin/sh" -c "tar czf /opt/bin/nvidia-bins.tar.gz /usr/bin/nvidia*"
+
+- name: send signal to terminate build container
+  command: "docker kill -s SIGHUP nvidia-driver"
+
+- name: extract libs into /opt/lib/x86_64-linux-gnu
+  command: "tar xzf /opt/lib/x86_64-linux-gnu.tar.gz --directory /opt/lib --strip-components 2"
+
+- name: remove x86_64-linux-gnu tarball
+  command: "rm /opt/lib/x86_64-linux-gnu.tar.gz"
+
+- name: extract bins into /opt/bin
+  command: "tar xzf /opt/bin/nvidia-bins.tar.gz --directory /opt/bin --strip-components 2"
+
+- name: remove nvidia-bins tarball
+  command: "rm /opt/bin/nvidia-bins.tar.gz"
+
+# FIXME(tillt): consider not packing them in the first place
+- name: kill dead libs in /opt/lib/x86_64-linux-gnu
```

does "kill" have a special meaning here, is that a term used somewhere else?
```suggestion
- name: remove dead libs in /opt/lib/x86_64-linux-gnu
```

---

The entire thing is very "sloppy". 

We are first getting a bunch of libraries from inside the container. Then we extract that tarball. The results will contain a bunch of broken links. Link destinations that existed within the container but not outside. Those have to get cleared to not completely get the system broken.

---

Added a comment describing the intend.